### ISSUE TITLE
Frequency Xlating FIR Filter does not generate reproducible results

### ISSUE THREAD

**Issue by reald** [gnuradio/gnuradio#6994]
### What happened?

While analyzing data with gnuradio I stumbled about slightly different results when rerunning my analysis the other day. 

At first I could find out that the **Frequency Xlating FIR Filter** does **produce different results** for the **same input data**. 

There might be more blocks generating non deterministic results.

### System Information

OS: Ubuntu 20.04.5 LTS
GR Installation Method: compile from source
CPU: AMD Ryzen 7 5700G with Radeon Graphics

### GNU Radio Version

3.11-git (main)

### Specific Version

0.0.0.0

### Steps to reproduce
1.  Generate test data: `dd if=/dev/urandom of=/tmp/test.cfile bs=100000000 count=1`
2.  Run test flow graph
![test](https://github.com/gnuradio/gnuradio/assets/1385404/7700083b-c1bf-4371-bb7e-a3878423249b)
3.  Compare files
```
md5sum /tmp/file*
860a0ea092969951aee3a93b06559ed1  /tmp/file1.cfile
1973c48093094ec10293e1e2b1ad2268  /tmp/file2.cfile`
```
The begin of both files is identical but then suddenly it begins to differ.

### Expected Behaviour
Same input data must produce same output data.

### Note
GR 3.10.1.1 in Ubuntu 22.04.3 LTS has the same issue.

**Comment by reald** [comment]
Flow graph for testing.

[test.zip](https://github.com/gnuradio/gnuradio/files/13603152/test.zip)

**Comment by willcode** [comment]
That `dd` and flowgraph produced 2 identical files for me. Linux, Fedora 38, latest GR maint-3.10 branch.

**Comment by willcode** [comment]
There were fixes to the rotator (used inside the freq xlating filter) for 3.10.3.0.

**Comment by reald** [comment]
Yeah, the test data might a bit short. Rerun it or increase the length. Does not take long to get different results.

**Comment by willcode** [comment]
Got a difference after a few tries.

**Comment by willcode** [comment]
Ah, bad test. The random number generator is probably producing invalid float values. Try using the random signal source instead, or at least use valid floats for the input. Reopen if the problem is confirmed with good input.

**Comment by reald** [comment]
No problem. Here is another option to generate input data. 

![gen_data](https://github.com/gnuradio/gnuradio/assets/1385404/a0e6da52-f147-48c1-939f-c0e2082792c6)

Let it run for a few seconds.

```
md5sum file*
b37ec5c62e267bef210e3a83978cd527  file1.cfile
e648aa019dfb1c9de95900e39eaa1b94  file2.cfile
```


Effect also happended with real measurements...

Flow graph in .zip.

[gen_data.zip](https://github.com/gnuradio/gnuradio/files/13603365/gen_data.zip)

**Comment by willcode** [comment]
Reproduced. I added a scale of 32768 to the short-to-float to make sure values were sane. In any case, I can't explain the difference and agree that there shouldn't be any.

Do you see any bad effects, like a dropped sample? Do the differences look like rounding/precision errors, or is the signal totally corrupted when this happens? It's hard to tell using random data.

**Comment by reald** [comment]
I stumbled across this after the same input data had a different number of packets with correct CRC. So I placed some file sinks into my flowgraph and the data after the Xlating FIR Filter was not the same after several tries. This was my first catch, there might be more.

I tried building the difference of 2 files and triggering on the magnitude of the difference:
![delta](https://github.com/gnuradio/gnuradio/assets/1385404/3aa48905-2310-467f-83ae-6cb94efb3fec)

It showed a delta in magnitude of around 0.5:

![Screenshot_20231207_192234](https://github.com/gnuradio/gnuradio/assets/1385404/9da428a5-fbda-4d24-9135-78159b798c2b)

Now I used a signal source instead of a random file:

![test3](https://github.com/gnuradio/gnuradio/assets/1385404/c8826242-ff2b-4220-8923-44532539af16)

This has the disavantage, that when stopping the flowgraph a different number of bytes is written to both files so the checksum method is not possible. But comparing them with `vbindiff` shows following picture:

![Screenshot_20231207_193950](https://github.com/gnuradio/gnuradio/assets/1385404/9907c650-d50e-4885-9379-af968de2d3d3)

It seems that the difference can reach the second byte which is a lot. Surprisingly after a time both signals get the same again... 

[flowgr.zip](https://github.com/gnuradio/gnuradio/files/13603775/flowgr.zip)

**Comment by willcode** [comment]
That's really interesting - thanks for diving into it. Don't know what to make of it yet. @argilo @marcusmueller any ideas?

**Comment by argilo** [comment]
I would guess it's due to the fact that the thread-per-block scheduler is not deterministic, and so different input chunk sizes can be passed into the rotator, causing it to re-normalize at different points.

**Comment by argilo** [comment]
> Is it not happening often enough?

It happens at least once every 512 samples:

https://github.com/gnuradio/volk/blob/d1b1673eb61f936cecee769255203be790b1ad27/kernels/volk/volk_32fc_s32fc_x2_rotator2_32fc.h#L83-L107

**Comment by argilo** [comment]
The issue is that `volk_32fc_s32fc_x2_rotator2_32fc` only knows about the samples in the current chunk, and so it normalizes every 512 samples from the first sample in the chunk (plus once more at the end). It has no idea how many samples came before. So you'll get different results in you break the input samples up into different chunk lengths.

**Comment by willcode** [comment]
@reald When a different number of packets had the right CRC, was the signal weak (would be on the edge of decoding correctly) or really strong?

**Comment by argilo** [comment]
As a hack, it might be possible to force deterministic behaviour by setting the filter block's output multiple to 512.

**Comment by argilo** [comment]
> The signal was strong but maybe not suitable for the symbol sync block.

It can be tricky to get Symbol Sync working reliably. Not all timing error detectors are suitable for all signal types, and the parameters may need to be adjusted to achieve quick synchronization if you're dealing with short packets. Another gotcha I sometimes run into is that Symbol Sync doesn't seem to work well if Samples per Symbol is set to a large value (e.g. > 20).

> Where can i set the blocks output?

I don't believe it's possible to do it within GNU Radio Companion. You'd have to edit the generated Python code and add something like `self.<my_freq_xlating_block_name>.set_output_multiple(512)` to the `# Blocks` section.

**Comment by argilo** [comment]
Agreed. My motivation for suggesting `set_output_multiple(512)` was to check whether the rotator's phase accumulator normalization was the source of the nondeterministic behaviour.

**Comment by ffu** [comment]
I tested on 3.10.12. I confirm setting the output multiple to 512 does indeed eliminate the non-deterministic behavior and creates reproducible results.

Next, I did some further analysis on the error between two xlating fir filter blocks. One important metric not discussed is relative error. For example, if one block outputs 1.5 and the other block outputs 1.0, that error of 0.5 is drastic. However, if the values are 20000 and 20000.5, the error of 0.5 appears slight. I created a flowgraph of a random source connected to two xlating blocks which then fed to a relative error calculation abs(a - b)/abs(b). I rarely see relative error values above 1.2e-5. From this, it seems like this wouldn't be a huge issue even if no fix was made.
